src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs

```diff
@@ -246,15 +262,27 @@ private List<ExtractedResult> ExtractResultsFromRuns(IEnumerable<Run> sarifRuns)
                 indexRemappingVisitor.HistoricalLogicalLocations = resultPair.Run.LogicalLocations;
                 indexRemappingVisitor.VisitResult(result);
 
+                // This logic is incomplete. It doesn't account for the possibility that result.rule
+                // is present (in which case the rule might come from an extension rather than from
+                // the driver).
                 if (result.RuleIndex != -1)
                 {
-                    ReportingDescriptor rule = resultPair.Run.Tool.Driver.Rules[0];
-                    if (!reportingDescriptors.TryGetValue(rule, out int ruleIndex))
+                    ReportingDescriptor rule = resultPair.Run.Tool.Driver.Rules.FirstOrDefault(r => r.Id.Equals(result.RuleId, StringComparison.Ordinal));
```

Comment:
>ReportingDescriptor [](start = 20, length = 19)

Is there an opportunity here to leverage the Result.GetRule(Run) method that is added to our Result object? See Core\Result.cs to assess. Rule look-up isn't easy. Would be great to keep expanding a single utility for this operation.

Reply:
Perfect! I didn't know/remember that one. It also means I can remove the comment above about the "incomplete logic" -- the logic is as complete as the helper is.

---
In reply to: [328779954](https://github.com/microsoft/sarif-sdk/pull/1686#discussion_r328779954) [](ancestors = 328779954)

Reply:
Done.

---
In reply to: [328781881](https://github.com/microsoft/sarif-sdk/pull/1686#discussion_r328781881) [](ancestors = 328781881,328779954)